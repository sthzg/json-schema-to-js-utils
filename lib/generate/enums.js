const { createWriteStream } = require('fs');
const {
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  FilterEnumsTransform,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  StdoutWritable,
  PrependToFileTransform,
  PrettierTransform,
} = require('../streams');
const { enumModule } = require('../templates');

module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  jsonSchemaReadable(schema)
    .pipe(new FilterPropertiesTransform())
    .pipe(new FilterEnumsTransform())
    .pipe(new BuildObjectPathsTransform())
    .pipe(new CompileToTemplateTransform(enumModule))
    .pipe(new MemorySinkTransform())
    .pipe(new PrependToFileTransform(DEFAULT_FILE_DOCSTRING))
    .pipe(new PrettierTransform(DEFAULT_PRETTIER_OPTIONS))
    .pipe(write);
};
