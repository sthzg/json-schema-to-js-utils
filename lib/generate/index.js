import generateEnumsModule from './enums';
import generateObjectPathsModule from './paths';
import generateSelectorsModule from './selectors';

export {
  generateEnumsModule,
  generateObjectPathsModule,
  generateSelectorsModule,
};
