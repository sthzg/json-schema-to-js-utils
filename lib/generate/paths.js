const { createWriteStream } = require('fs');
const { pipeline: pipelineFn } = require('stream');
const { promisify } = require('util');
const {
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  StdoutWritable,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  PrependToFileTransform,
  PrettierTransform,
} = require('../streams');
const { objectPathConstant } = require('../templates');

const pipeline = promisify(pipelineFn);

module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  return pipeline(
    jsonSchemaReadable(schema),
    new FilterPropertiesTransform(),
    new BuildObjectPathsTransform(),
    new CompileToTemplateTransform(objectPathConstant),
    new MemorySinkTransform(),
    new PrependToFileTransform(DEFAULT_FILE_DOCSTRING),
    new PrettierTransform(DEFAULT_PRETTIER_OPTIONS),
    write
  );
};
