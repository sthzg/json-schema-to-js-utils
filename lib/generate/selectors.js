const { createWriteStream } = require('fs');
const { pipeline: pipelineFn } = require('stream');
const { promisify } = require('util');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  StdoutWritable,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  PrettierTransform,
  PrependToFileTransform,
} = require('../streams');
const {
  DEFAULT_GET_IMPORT,
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const { selectorFunctionModule } = require('../templates');

const pipeline = promisify(pipelineFn);

/**
 * Generates selector functions for all values inside the JSON schema.
 *
 * Selectors are functions that pick a value from an object in a failsafe way or return undefined if
 * the property is not defined on the input object, e.g.
 *
 *     function byPersonAddressStreet(person) {
 *       return get(person, 'address.street');
 *     }
 */
module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  return pipeline(
    jsonSchemaReadable(schema),
    new FilterPropertiesTransform(),
    new BuildObjectPathsTransform(),
    new CompileToTemplateTransform(selectorFunctionModule),
    new MemorySinkTransform(),
    new PrependToFileTransform(DEFAULT_FILE_DOCSTRING, DEFAULT_GET_IMPORT),
    new PrettierTransform(DEFAULT_PRETTIER_OPTIONS),
    write
  );
};
