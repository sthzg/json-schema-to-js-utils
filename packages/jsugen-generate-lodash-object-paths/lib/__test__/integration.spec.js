import schema from '@sthzg/jsugen-core/lib/testResources/example.schema';
import generate from '../index';

test('generate-lodash-object-paths integration test', () => {
  expect.assertions(1);

  generate({ schema }).subscribe(output => {
    expect(output).toMatchSnapshot();
  });
});
