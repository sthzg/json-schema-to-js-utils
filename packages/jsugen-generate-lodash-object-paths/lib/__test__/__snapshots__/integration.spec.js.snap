// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-lodash-object-paths should generate expected paths for array-of-objects properties 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byRootNthLevelOneNthFooPath = (
  rootIndex = 0,
  levelOneIndex = 0,
) => ['root', rootIndex, 'levelOne', levelOneIndex, 'foo'];

export const byRootNthLevelOnePath = (rootIndex = 0) => [
  'root',
  rootIndex,
  'levelOne',
];

export const byRootNthLevelOneNthPath = (rootIndex = 0, levelOneIndex = 0) => [
  'root',
  rootIndex,
  'levelOne',
  levelOneIndex,
];

export const byRootNthLevelOneWithoutItemsDefPath = (rootIndex = 0) => [
  'root',
  rootIndex,
  'levelOneWithoutItemsDef',
];

export const byRootNthLevelOneWithoutItemsDefNthPath = (
  rootIndex = 0,
  levelOneWithoutItemsDefIndex = 0,
) => [
  'root',
  rootIndex,
  'levelOneWithoutItemsDef',
  levelOneWithoutItemsDefIndex,
];

export const byRootNthAnyMemberPath = (rootIndex = 0) => [
  'root',
  rootIndex,
  'anyMember',
];

export const byRootPath = () => ['root'];

export const byRootNthPath = (rootIndex = 0) => ['root', rootIndex];
"
`;

exports[`generate-lodash-object-paths should pass example.contact.json integration test 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byIdPath = () => ['id'];

export const byFirstNamePath = () => ['firstName'];

export const byLastNamePath = () => ['lastName'];

export const byGenderPath = () => ['gender'];

export const byPhoneNthTypePath = (phoneIndex = 0) => [
  'phone',
  phoneIndex,
  'type',
];

export const byPhoneNthNumberPath = (phoneIndex = 0) => [
  'phone',
  phoneIndex,
  'number',
];

export const byPhonePath = () => ['phone'];

export const byPhoneNthPath = (phoneIndex = 0) => ['phone', phoneIndex];
"
`;

exports[`generate-lodash-object-paths should pass full integration test 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byLevel1Member1Path = () => ['level1Member1'];

export const byLevel1Member2Path = () => ['level1Member2'];

export const byLevel1Member3Path = () => ['level1Member3'];

export const byLevel1Member4Path = () => ['level1Member4'];

export const byLevel1Member5Path = () => ['level1Member5'];

export const byLevel1Member5NthPath = (level1Member5Index = 0) => [
  'level1Member5',
  level1Member5Index,
];

export const byLevel1Member6NthLevel2Member1Path = (level1Member6Index = 0) => [
  'level1Member6',
  level1Member6Index,
  'level2Member1',
];

export const byLevel1Member6Path = () => ['level1Member6'];

export const byLevel1Member6NthPath = (level1Member6Index = 0) => [
  'level1Member6',
  level1Member6Index,
];

export const byLevel1Member7DefinitionMember1Path = () => [
  'level1Member7',
  'definitionMember1',
];

export const byLevel1Member7Path = () => ['level1Member7'];

export const byLevel1Member8Level2Member1Path = () => [
  'level1Member8',
  'level2Member1',
];

export const byLevel1Member8Level2Member2Path = () => [
  'level1Member8',
  'level2Member2',
];

export const byLevel1Member8Level2Member3Path = () => [
  'level1Member8',
  'level2Member3',
];

export const byLevel1Member8Level2Member3NthPath = (level2Member3Index = 0) => [
  'level1Member8',
  'level2Member3',
  level2Member3Index,
];

export const byLevel1Member8Level2Member4Path = () => [
  'level1Member8',
  'level2Member4',
];

export const byLevel1Member8Level2Member5Path = () => [
  'level1Member8',
  'level2Member5',
];

export const byLevel1Member8Level2Member6Level3Member1Path = () => [
  'level1Member8',
  'level2Member6',
  'level3Member1',
];

export const byLevel1Member8Level2Member6Path = () => [
  'level1Member8',
  'level2Member6',
];

export const byLevel1Member8Path = () => ['level1Member8'];

export const byLevel1Member9Path = () => ['level1Member9'];
"
`;
